README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# sjofun

<!-- badges: start -->

[![Codecov test
coverage](https://codecov.io/gh/ddsjoberg/sjofun/branch/master/graph/badge.svg)](https://codecov.io/gh/ddsjoberg/sjofun?branch=master)
[![R build
status](https://github.com/ddsjoberg/sjofun/workflows/R-CMD-check/badge.svg)](https://github.com/ddsjoberg/sjofun/actions)
<!-- badges: end -->

I collection of functions I’ve written to assist with my projects…

## Installation

You can install sjofun from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("ddsjoberg/sjofun")
```


NEWS.md
# sjofun 0.1.0

* Added a `NEWS.md` file to track changes to the package.


R/dlogitnorm.R
#' Logit-normal distribution functions
#'
#' The logit-normal distribution is a probability distribution of a random variable
#' whose logit has a normal distribution. If Y is a random variable with a normal
#' distribution, and P is the logistic function, then X = P(Y) has a logit-normal
#' distribution; likewise, if X is logit-normally distributed, then
#' Y = logit(X)= log (X/(1-X)) is normally distributed.
#'
#' @author Daniel D Sjoberg \email{sjobergd@@mskcc.org}
#'
#' @param x input in [0,1]
#' @param n number of random variables to generate
#' @param location is the location parameter
#' @param scale is the scale parameter
#'
#' @return Returns a list of results from analysis.
#'
#' @examples
#' x=seq(0,1,0.1)
#' y=dlogitnorm(x)
#' plot(x,y)
#'
#'
#' @export
#'
# density function
dlogitnorm=function(x, location = 0, scale = 1){
  #function doesn't exist at 0 and 1, os making it zero
  ifelse(0<x & x<1,
         dnorm(qlogis(x), mean = location , sd = scale)*((x*(1-x))^-1),
         0)
}
# CDF
plogitnorm=function(x, location = 0, scale = 1){
  pnorm(qlogis(x), mean = location, sd = scale)
}
# random var generator
rlogitnorm=function(n, location = 0, scale = 1){
  exp(rnorm(qlogis(x), mean = location, sd = scale))
}


R/model.diff.R
#' Model-based test for treatment efficacy
#'
#' The model-based sinlge-arm comparison addresses the fundamental question of whether the current treatment provides a clinically significant
#' improvement over prior treatments in the population. The proposed test statistic computes the difference between the
#' observed outcome from the current treatment and the covariate-specific predicted outcome based on a model of the
#' historical data. Thus, the difference between the observed and predicted quantities is attributed to the current
#' treatment.
#'
#' Heller, Glenn, Michael W. Kattan, and Howard I. Scher. "Improving the decision to pursue a phase 3 clinical trial
#' by adjusting for patient-specific factors in evaluating phase 2 treatment efficacy data." Medical Decision Making
#' 27.4 (2007): 380-386.
#'
#' @author Daniel D Sjoberg \email{sjobergd@@mskcc.org}
#'
#' @param data a data frame containing the outcome and the outcome predictions.
#' @param model glm object of the predictive model estimated on historical cohort.  If specified, the outcome, covars, cov, and coef objects will be extracted from object.
#' @param outcome the outcome, or response variable name. Must be a variable contained within the data frame specified in data=.
#' @param covars vector of covariate/predictor variable(s) names. Must be a variable(s) contained within the data frame specified in data=.  If model includes an intercept, user must include the column of ones in the covars vector
#' @param cov Variance-covaraince matrix of the beta coefficients from predictive model.
#' @param coef Vector of the beta coefficients from predictive model.  If model includes an intercept, a vector of ones must appear in data.
#' @param type Type of predictive model used.  logistic is currently the only valid input.
#' @param output.details Save additional information.  Default is FALSE.
#'
#' @return Returns a list of results from analysis.
#'
#' @examples
#' set.seed(23432)
#' #simulating historic dataset and creating prediction model.
#' marker=rnorm(500, sd = 2)
#' respond=runif(500)<plogis(marker)
#' historic.data=data.frame(respond,marker)
#' model.fit=glm(data=historic.data, formula = respond ~ marker, family = binomial(logit))
#'
#' #simulating new data, with higher response rate
#' new.data = marker=rnorm(50, sd = 2)
#' respond=runif(50)<plogis(marker + 1)
#' new.data=data.frame(respond,marker)
#'
#' #comparing outcomes in new data to those predicted in historic data
#' # z-statistic = 2.412611 indicates signficant difference
#' model_diff(data = new.data, model = model.fit)
#'
#' #comparing model based difference with binomial test
#' #p-value of 0.3222 indicates we fail to reject null hypothesis
#' binom.test(x=sum(new.data$respond), n=nrow(new.data), p = 0.5, alternative = c("two.sided"))
#'
#' @export

model_diff=function(data, outcome, covars, model, cov, coef, type="logistic", output.details = FALSE) {
  #initializing empty list to store results
  results=list()

  #checking to assess whether prior model was provided.
  if(missing(model) & (missing(cov) | missing(coef)))  {
    stop("Must provide model object (model), or both covariance matrix (cov) and coefficient vector (coef).")
  } else if(!missing(model) & (!missing(cov) | !missing(coef)) ) {
    stop("Provide model object (model), or covariance matrix (cov) and coefficient vector (coef).  But not both.")
  }


  #extracting covariance and coefficient if model object provided, calculating linear predictor
  if(!missing(model)){
    #extracting linear predictor
    pr.xb=stats::predict(model, newdata = data)
    #extracting variance-covariance matrix
    cov=stats::vcov(model)
    #extracting vector of coefficients
    coef=model$coefficients
    #extracting covaraite names
    covars=names(model$coefficients)
    #if model includes intercept, adding vector of ones to data
    if(covars[1]=="(Intercept)") data["(Intercept)"]=1
    #extracting outcome name
    outcome=all.vars(attributes(model$terms)$variables)[1]
  }

  #subsetting on complete data
  data=data[stats::complete.cases(data[,c(outcome, covars)]),]

  #if the model object not specified, using use supplied info to calculate the XB
  if(missing(model)) pr.xb=data[,covars] %*% coef


  #calculating difference in reposnse and predicted outcome
  pr.outcome=exp(pr.xb)/(1+exp(pr.xb))
  delta=data[,outcome] - pr.outcome

  #saving observation level data
  if(output.details==TRUE){
    results[[1]]=cbind(data[,c(outcome,covars)],pr.outcome,pr.xb,delta)
    names(results[[1]])[1:length(c(outcome,covars))]=c(outcome,covars)
    names(results)=c("data")
  }
  #calculating variance of Y- pr(y)
  V1=mean(delta^2)
  V2=cov
  d=colMeans(apply(data[,covars], 2, function(x) x*pr.outcome*(1-pr.outcome)))
  V=as.numeric(nrow(data)*(V1 + t(d) %*% V2 %*% d))
  var.list=list(V, V1, V2, d)
  names(var.list)=c("var.tot","V1","V2","d")
  results=c(results,var.list)

  #calculating test statistic
  S.star=as.vector(sum(delta)/sqrt(V))
  results=c(results,S.star)
  names(results)[length(results)]="z"

  #returning final result
  return(results)
}


R/density.auc.R
#' Calculate exact AUCs based on the riskribution of risk in a population.
#'
#' Provided a distribution of risks in a population, this function calculates the
#' exact AUC of a model that produces the risk estiamtes.  For example, a logistic
#' regression model built with a normal linear predictor yields logit-normal distributed
#' predicted risks.  The AUC from the logisitc regression model is the same as the AUC estimated
#' from the distribution of the predicted risks, indpendent of the outcome.  This method for AUC
#' calculation is useful for simulative studies where the predicted risks are a mixture of two
#' distributions.  The exact prevelance of the outcome can easily be calculated, along with the exact
#' AUC of the model.
#'
#' @author Daniel D Sjoberg \email{sjobergd@@mskcc.org}
#'
#' @param density a function name that descirbes the continuous probability density function of the
#' risk from 0 to 1.
#' @param cut.points sequence of points in [0, 1] where the sensitivity and specificity are calculated.
#' More points lead to a more precise estimate of the AUC.  Default is seq(from = 0, to = 1,by = 0.001).
#' @param ... arguments for the function specified in density.  For example, dbeta(x, shape1=1, shape2=1)
#' has need for two additional arguments to specify the density function (shape1 and shape2).
#'
#' @return Returns a list sensitities and specifcities at each cut point, the expected value or
#' mean risk, and the AUC associated with the distribution.
#'
#' @examples
#' density.auc(density=dbeta, shape1=1, shape2=1)
#'
#' @export
#'
density.auc=function(density, cut.points=seq(from = 0, to = 1,by = 0.001), ...){
  # calculating distribution mean
  mu=integrate(function(x) x*density(x, ...) ,0,1)$value

  sensitivity=NULL
  specificity=NULL
  for(c in cut.points){
    #calculating Sens and Spec using Bayes Rule
    sens0=integrate(function(x) x*density(x, ...), c, 1)$value / mu
    spec0=(integrate(function(x) density(x, ...) ,0,c)$value -
             integrate(function(x) x*density(x, ...) ,0,c)$value) /
      (1 - mu)

    #appending calculated results
    sensitivity=c(sensitivity,sens0)
    specificity=c(specificity,spec0)
  }

  #calculating the AUC (using the trapezoidal rule)
  idx = 2:length(cut.points)
  auc=as.numeric(-(specificity[idx-1] - specificity[idx]) %*% (sensitivity[idx] + sensitivity[idx-1]) / 2)

  #retruning results
  results=list(cut.points, sensitivity, specificity, mu, auc)
  names(results)=c("cut.point", "sensitivity", "specificity", "mu", "auc")

  return(results)
}

density.auc(density=dbeta, shape1=1, shape2=1)


R/golden_search.R
#' One-diminsional Search Algorithm
#'
#' Find the minimum of a function using the Golded-section search algorithm
#'
#' @param x function with one argument
#' @param a min value of search values
#' @param b max value of search values
#' @param tol tolerance level. Default is `1e-04`
#'
#' @export
#' @examples
#' contfunc = function(x){
#'   -exp(-x) * sin(x)
#' }
#' golden_search(contfunc, a = 0, b = 1.5)
golden_search <- function(x, a, b, tol = 1e-04) {
  phi <- (-1 + sqrt(5)) / 2

  c <- b - phi * (b - a)
  d <- a + phi * (b - a)
  i <- 0
  iter <- NULL
  while (abs(c - d) > tol) {
    i <- i + 1

    fc <- x(c)
    fd <- x(d)

    # to find min, use < ; > is for max
    if (fc < fd) {
      b <- d
      d <- c
      # fd=fc;fc=f(c)
      c <- b - phi * (b - a)
    } else {
      a <- c
      c <- d
      # fc=fd;fd=f(d)
      d <- a + phi * (b - a)
    }

  }
  return((b + a) / 2)
}


R/ph2_single_bayes_sim.R
#' Simulate results from a single arm Bayesian Phase 2 Trial
#'
#' Trial is a Bayesian phase 2, sequential trial as outlined in
#' Thall, Peter F., and Richard Simon. "Practical Bayesian guidelines for phase
#' IIB clinical trials." Biometrics (1994): 337-349.
#'
#' @param sim_n Number of simulated trials to create
#' @param omega_s_mu Mean of prior distribution on omega_s
#' @param omega_s_width Width of the probability interval (set by `w_conf_level=`)
#' running from the `(1 + w_conf_level) / 2` to the `(1 - w_conf_level) / 2` percentiles
#' @param omega_e_c Concentration parameter for prior distribution on omega_e
#' @param delta_0 Targeted improvement of treatment for treatment E over S
#' @param n_min Minimum number of patients that may be enrolled
#' @param n_max Maximum number of patients that may be enrolled
#' @param pr_low Lower probability limit for concluding treatment not promising
#' @param pr_high Upper probablity limit for concluding treatment promising
#' @param omega_s_w_conf_level Confidence level of width of probabilty specified
#' @param mu_e True rate of success with experimental treatment
#' in `omega_s_width`
#' @param verbose When TRUE, additional information is returned as an attribute
#' @param quiet Run with no notes, progress bars, etc.
#' @export

ph2_single_bayes_seq_sim <- function(omega_s_mu, omega_s_width,
                                     omega_s_w_conf_level = 0.90,
                                     omega_e_c, delta_0, n_min, n_max,
                                     sim_n = 1, pr_low = 0.05, pr_high = 0.95,
                                     mu_e = omega_s_mu + delta_0,
                                     verbose = FALSE, quiet = FALSE) {
  # input checks ---------------------------------------------------------------

  # prior distributions --------------------------------------------------------
  # getting parameters for prior distribution of standard treatment
  prior_omega_s <-
    cnvrt_w_to_beta_param(mu = omega_s_mu, w = omega_s_width,
                          w_conf_level = omega_s_w_conf_level)

  # getting parameters for prior distribution of experimental treatment
  prior_omega_e <- list(
    shape1 = omega_e_c * (omega_s_mu + delta_0 / 2),
    shape2 = omega_e_c * (1 - (omega_s_mu + delta_0 / 2) )
  )

  # simulating trial results ---------------------------------------------------
  if (quiet == FALSE) pb <- dplyr::progress_estimated(sim_n) # this is the progress bar
  result <-
    seq_len(sim_n) %>%
    map_dfr(
      function(.x) {
        if (quiet == FALSE) pb$tick()$print()
        smry_trial_result(n_min = n_min, n_max = n_max, mu_e = mu_e,
                        delta_0 = delta_0,
                        prior_omega_s = prior_omega_s,
                        prior_omega_e = prior_omega_e,
                        pr_low = pr_low, pr_high = pr_high) %>%
        as_tibble() %>%
        mutate(sim_id = .x) %>%
        select(.data$sim_id, everything())
      }
    )

  # including additional results if requested
  if (verbose == TRUE) {
    attr(result, "prior_omega_s") <- prior_omega_s
    attr(result, "prior_omega_e") <- prior_omega_e
  }

  result
}






# helper function to summarize trial results
smry_trial_result <- function(n_min, n_max, mu_e,
                              delta_0,
                              prior_omega_s,
                              prior_omega_e,
                              pr_low, pr_high) {


  df_trial_outcome <-
    tibble(
      id = seq_len(n_max),
      outcome = as.numeric(stats::runif(n_max) < mu_e),
      cum_sucess = cumsum(.data$outcome)
    ) %>%
    # decisions only made after n_min enrolled
    filter(.data$id >= n_min) %>%
    mutate(
      lambda_prob = map2_dbl(
        .data$id, .data$cum_sucess,
        ~lambda_fun(n = .x, x = .y, delta_0, prior_omega_s, prior_omega_e)
      ),
      stop_futility = .data$lambda_prob <= .env$pr_low,
      stop_success = .data$lambda_prob >= .env$pr_high
    )


  # returning trial results ----------------------------------------------------
  stop_futility <-
    which(df_trial_outcome$stop_futility) %>%
    {suppressWarnings(min(.))} %>%
    discard(is.infinite)
  stop_success <- which(df_trial_outcome$stop_success) %>%
    {suppressWarnings(min(.))} %>%
    discard(is.infinite)

  # inconclusive
  if (length(stop_futility) == 0 && length(stop_success) == 0) {
    return(list(
      result = "inconclusive",
      n_enrolled = n_max
    ))
  }
  # success
  if (
    (length(stop_futility) == 0 && length(stop_success) > 0) ||
    (length(stop_futility) > 0 && length(stop_success) > 0 && stop_success < stop_futility)
  ) {
    return(list(
      result = "success",
      n_enrolled = df_trial_outcome$id[stop_success]
    ))
  }
  # futility
  if (
    (length(stop_success) == 0 && length(stop_futility) > 0) ||
    (length(stop_success) > 0 && length(stop_futility) > 0 && stop_futility < stop_success)
  ) {
    return(list(
      result = "futile",
      n_enrolled = df_trial_outcome$id[stop_futility]
    ))
  }

  # there should be nothing here
  stop("There was an error")
}

# this is the lambda function from the paper that defines probs of success, futility
lambda_fun <- function(n, x, # number of obs = n, x = successes
                       delta_0, prior_omega_s, prior_omega_e) {
  inside_fun <- function(p) {
    cdf_portion <- stats::pbeta(p + delta_0,
                                shape1 = prior_omega_e$shape1 + x,
                                shape2 = prior_omega_e$shape2 + n - x)

    pdf_portion <- stats::dbeta(p,
                                shape1 = prior_omega_s$shape1,
                                shape2 = prior_omega_s$shape2)

    (1 - cdf_portion) * pdf_portion
  }

  stats::integrate(inside_fun, lower = 0, upper = 1 - delta_0)$value
}

# helper function that converts mu and w into a and b Beta parameterizations
cnvrt_w_to_beta_param <- function(w, w_conf_level, mu) {
  # this function takes an x in (0, 1), converts it to a in (0, Inf),
  # calculates the w, and returns teh difference between the calculated w
  # and our target w
  find_beta_params <- function(x, w, w_conf_level) {
    a = -log(x)
    b = a * (1 - mu) / mu

    w_est <-
      stats::qbeta((1 + w_conf_level) / 2, shape1 = a, shape2 = b) -
      stats::qbeta((1 - w_conf_level) / 2, shape1 = a, shape2 = b)

    abs(w - w_est)
  }

  # this finds the x that results in the w closest our target
  x <- golden_search(
    function(x) find_beta_params(x, w = w, w_conf_level = w_conf_level),
    a = 0, b = 1, tol = 1e-10
  )

  # return the a and b parameters (or the shape1 and shape2)
  list(
    shape1 = -log(x),
    shape2 = -log(x) * (1 - mu) / mu
  )
}




R/ph2_single_bayes_seq_sim.R
#' Simulate results from a single arm Bayesian Phase 2 Trial
#'
#' Trial is a single arm Bayesian phase 2 trial with sequential stopping boundaries
#'  as outlined in
#' Thall, Peter F., and Richard Simon. "Practical Bayesian guidelines for phase
#' IIB clinical trials." Biometrics (1994): 337-349.
#'
#' @param sim_n Number of simulated trials to create
#' @param theta_s_mu Mean of prior distribution on theta_s
#' @param theta_s_width Width of the probability interval (set by `w_conf_level=`)
#' running from the `(1 + w_conf_level) / 2` to the `(1 - w_conf_level) / 2` percentiles
#' @param theta_e_c Concentration parameter for prior distribution on theta_e
#' @param delta_0 Targeted improvement of treatment for treatment E over S
#' @param n_min Minimum number of patients that may be enrolled
#' @param n_max Maximum number of patients that may be enrolled
#' @param pr_low Lower probability limit for concluding treatment not promising
#' @param pr_high Upper probablity limit for concluding treatment promising
#' @param theta_s_w_conf_level Confidence level of width of probabilty specified
#' @param mu_e True rate of success with experimental treatment
#' in `theta_s_width`
#' @param verbose When TRUE, additional information is returned as an attribute
#' @param quiet Run with no notes, progress bars, etc.
#' @export
#'
#' @examples
#' # simulate trial results
#' sim_results <- ph2_single_bayes_seq_sim(
#'   # setting priors for standard treatment
#'   theta_s_mu = 0.2, theta_s_width = 0.20, theta_s_w_conf_level = 0.90,
#'   # setting priors for experimental treatment
#'   theta_e_c = 2, delta_0 = 0.15,
#'   # other trial parameters
#'   n_min = 10, n_max = 65,
#'   pr_low = 0.05, pr_high = 0.95,
#'   # true effect of experimental tx
#'   mu_e = 0.35,
#'   # number of simulations
#'   sim_n = 1000
#' )
#'
#' # tabulate summary
#' library(gtsummary)
#' sim_results %>%
#'   dplyr::select(-sim_id) %>%
#'   tbl_summary(
#'     label = list(result ~ "Trial Result",
#'                  n_enrolled ~ "No. Enrolled in Trial")
#'   ) %>%
#'   add_stat_label() %>%
#'   as_kable()

ph2_single_bayes_seq_sim <- function(theta_s_mu, theta_s_width,
                                     theta_s_w_conf_level = 0.90,
                                     theta_e_c, delta_0, n_min, n_max,
                                     sim_n = 1, pr_low = 0.05, pr_high = 0.95,
                                     mu_e = theta_s_mu + delta_0,
                                     verbose = FALSE, quiet = FALSE) {
  # input checks ---------------------------------------------------------------

  # prior distributions --------------------------------------------------------
  # getting parameters for prior distribution of standard treatment
  prior_theta_s <-
    cnvrt_w_to_beta_param(mu = theta_s_mu, w = theta_s_width,
                          w_conf_level = theta_s_w_conf_level)

  # getting parameters for prior distribution of experimental treatment
  prior_theta_e <- list(
    shape1 = theta_e_c * (theta_s_mu + delta_0 / 2),
    shape2 = theta_e_c * (1 - (theta_s_mu + delta_0 / 2) )
  )

  # simulating trial results ---------------------------------------------------
  if (quiet == FALSE) pb <- dplyr::progress_estimated(sim_n) # this is the progress bar
  result <-
    seq_len(sim_n) %>%
    map_dfr(
      function(.x) {
        if (quiet == FALSE) pb$tick()$print()
        smry_trial_result(n_min = n_min, n_max = n_max, mu_e = mu_e,
                        delta_0 = delta_0,
                        prior_theta_s = prior_theta_s,
                        prior_theta_e = prior_theta_e,
                        pr_low = pr_low, pr_high = pr_high) %>%
        as_tibble() %>%
        mutate(sim_id = .x) %>%
        select(.data$sim_id, everything())
      }
    )

  # including additional results if requested
  if (verbose == TRUE) {
    attr(result, "prior_theta_s") <- prior_theta_s
    attr(result, "prior_theta_e") <- prior_theta_e
  }

  result
}






# helper function to summarize trial results
smry_trial_result <- function(n_min, n_max, mu_e,
                              delta_0,
                              prior_theta_s,
                              prior_theta_e,
                              pr_low, pr_high) {


  df_trial_outcome <-
    tibble(
      id = seq_len(n_max),
      outcome = as.numeric(stats::runif(n_max) < mu_e),
      cum_sucess = cumsum(.data$outcome)
    ) %>%
    # decisions only made after n_min enrolled
    filter(.data$id >= n_min) %>%
    mutate(
      lambda_prob = map2_dbl(
        .data$id, .data$cum_sucess,
        ~lambda_fun(n = .x, x = .y, delta_0, prior_theta_s, prior_theta_e)
      ),
      stop_futility = .data$lambda_prob <= .env$pr_low,
      stop_success = .data$lambda_prob >= .env$pr_high
    )


  # returning trial results ----------------------------------------------------
  stop_futility <-
    which(df_trial_outcome$stop_futility) %>%
    {suppressWarnings(min(.))} %>%
    discard(is.infinite)
  stop_success <- which(df_trial_outcome$stop_success) %>%
    {suppressWarnings(min(.))} %>%
    discard(is.infinite)

  # inconclusive
  if (length(stop_futility) == 0 && length(stop_success) == 0) {
    return(list(
      result = "inconclusive",
      n_enrolled = n_max
    ))
  }
  # success
  if (
    (length(stop_futility) == 0 && length(stop_success) > 0) ||
    (length(stop_futility) > 0 && length(stop_success) > 0 && stop_success < stop_futility)
  ) {
    return(list(
      result = "success",
      n_enrolled = df_trial_outcome$id[stop_success]
    ))
  }
  # futility
  if (
    (length(stop_success) == 0 && length(stop_futility) > 0) ||
    (length(stop_success) > 0 && length(stop_futility) > 0 && stop_futility < stop_success)
  ) {
    return(list(
      result = "futile",
      n_enrolled = df_trial_outcome$id[stop_futility]
    ))
  }

  # there should be nothing here
  stop("There was an error")
}

# this is the lambda function from the paper that defines probs of success, futility
lambda_fun <- function(n, x, # number of obs = n, x = successes
                       delta_0, prior_theta_s, prior_theta_e) {
  inside_fun <- function(p) {
    experimental_tx_portion <-
      1 - stats::pbeta(p + delta_0,
                       shape1 = prior_theta_e$shape1 + x,
                       shape2 = prior_theta_e$shape2 + n - x)

    standard_tx_portion <-
      stats::dbeta(p,
                   shape1 = prior_theta_s$shape1,
                   shape2 = prior_theta_s$shape2)

    experimental_tx_portion * standard_tx_portion
  }

  stats::integrate(inside_fun, lower = 0, upper = 1 - delta_0)$value
}

# helper function that converts mu and w into a and b Beta parameterizations
cnvrt_w_to_beta_param <- function(w, w_conf_level, mu) {
  # this function takes an x in (0, 1), converts it to a in (0, Inf),
  # calculates the w, and returns teh difference between the calculated w
  # and our target w
  find_beta_params <- function(x, w, w_conf_level) {
    a = -log(x)
    b = a * (1 - mu) / mu

    w_est <-
      stats::qbeta((1 + w_conf_level) / 2, shape1 = a, shape2 = b) -
      stats::qbeta((1 - w_conf_level) / 2, shape1 = a, shape2 = b)

    abs(w - w_est)
  }

  # this finds the x that results in the w closest our target
  x <- golden_search(
    function(x) find_beta_params(x, w = w, w_conf_level = w_conf_level),
    a = 0, b = 1, tol = 1e-10
  )

  # return the a and b parameters (or the shape1 and shape2)
  list(
    shape1 = -log(x),
    shape2 = -log(x) * (1 - mu) / mu
  )
}


